crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -320,37 +413,25 @@ where
         Ok(())
     }
 
-    /// Pushes a new branch onto the `branch_stack`, while also pushing the given leaf onto the
-    /// `child_stack`.
+    /// Pushes a new branch onto the `branch_stack` based on the path and short key of the last
+    /// child on the `child_stack` and the path of the next child which will be pushed on to the
+    /// stack after this call.
     ///
-    /// This method expects that there already exists a child on the `child_stack`, and that that
-    /// child has a non-zero short key. The new branch is constructed based on the top child from
-    /// the `child_stack` and the given leaf.
-    fn push_new_branch(
-        &mut self,
-        targets: &mut TargetsIter<impl Iterator<Item = Nibbles>>,
-        leaf_key: Nibbles,
-        leaf_val: VE::DeferredEncoder,
-    ) -> Result<(), StateProofError> {
-        // First determine the new leaf's shortkey relative to the current branch. If there is no
-        // current branch then the short key is the full key.
-        let leaf_short_key = if self.branch_stack.is_empty() {
-            leaf_key
+    /// Returns the nibble of the branch's `state_mask` which should be set for the new child, and
+    /// short key that the next child should use.
+    fn push_new_branch(&mut self, new_child_path: Nibbles) -> (u8, Nibbles) {
```

Comment:
Previously this method was only used when pushing leafs onto the `child_stack`, now it is also used when pushing cached branches onto the `child_stack`. The core logic is the same, the only difference is that variables and documentation is made generic and `push_new_leaf` isn't called from here anymore; instead this method returns the nibble and short-key of the node which caused this new branch to be pushed, and the caller is expected to construct that node and push it onto the `child_stack` or `branch_stack` themselves.